### ISSUE TITLE
[Gap Analysis] Implement Dropout and Early Stopping Regularization

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#332]
### User Story

> As a machine learning engineer, I want access to a comprehensive set of regularization techniques, including Dropout and Early Stopping, so that I can effectively prevent overfitting and improve the generalization capabilities of my models, especially deep neural networks.

---

### **Problem: Missing Key Regularization Techniques**

The `src/Regularization` module currently provides L1, L2, and Elastic Regularization. However, two fundamental regularization techniques, widely used in deep learning and general machine learning, are missing.

---

### **Phase 1: Implement Dropout Regularization**

**Goal:** Add Dropout as a regularization technique for neural networks.

#### **AC 1.1: Create `DropoutRegularization.cs` (8 points)**
**Requirement:** Implement Dropout regularization.

-   [ ] **File:** `src/Regularization/DropoutRegularization.cs`
-   [ ] **Class:** `public class DropoutRegularization<T> : RegularizationBase<T>`
-   [ ] **Constructor:** Takes `double dropoutRate`.
-   [ ] **Methods:** `public override Matrix<T> Apply(Matrix<T> input, bool isTraining)`, `public override Matrix<T> Backward(Matrix<T> gradient)`.
-   [ ] **Logic:** Randomly sets a fraction of input units to zero during training; scales activations during inference.

#### **AC 1.2: Unit Tests for `DropoutRegularization` (5 points)**
**Requirement:** Verify the correctness of Dropout regularization.

-   [ ] **File:** `tests/UnitTests/Regularization/DropoutRegularizationTests.cs`
-   [ ] **Test Cases:** Test `Apply` in training mode (random zeros) and inference mode (scaling), and `Backward` pass.

---

### **Phase 2: Implement Early Stopping**

**Goal:** Add a mechanism to stop training when validation performance degrades.

#### **AC 2.1: Create `EarlyStopping.cs` (8 points)**
**Requirement:** Implement Early Stopping functionality.

-   [ ] **File:** `src/Regularization/EarlyStopping.cs`
-   [ ] **Class:** `public class EarlyStopping<T>`
-   [ ] **Constructor:** Takes `int patience`, `double minDelta`.
-   [ ] **Method:** `public bool ShouldStop(T currentValidationLoss)`: Returns true if training should stop.
-   [ ] **Logic:** Monitors validation loss and stops if it doesn't improve by `minDelta` for `patience` epochs.

#### **AC 2.2: Unit Tests for `EarlyStopping` (5 points)**
**Requirement:** Verify the correctness of Early Stopping.

-   [ ] **File:** `tests/UnitTests/Regularization/EarlyStoppingTests.cs`
-   [ ] **Test Cases:** Test scenarios where training should stop and where it should continue.

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] `DropoutRegularization` and `EarlyStopping` are implemented and unit-tested.
-   [ ] All new tests pass.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
